### PR TITLE
Add preserve nvvm

### PR DESCRIPTION
**Pull request by wsmoses**

### CODE REVIEW

**Comment by vchuravy** [comment]
Does this need a package bump?

**Comment by wsmoses** [comment]
No it does not, but its not complete yet.

File: src/compiler.jl [hunk 1]
```diff
@@ -1279,7 +1279,7 @@ function nested_codegen!(mod::LLVM.Module, f, tt)
     params = Compiler.PrimalCompilerParams()
     job    = CompilerJob(target, funcspec, params)
 
-    otherMod, meta = GPUCompiler.codegen(:llvm, job; optimize=false, validate=false, ctx)
+    otherMod, meta = GPUCompiler.codegen(:llvm, job; optimize=false, clean=false, validate=false, ctx)
```
Comment (vchuravy) [inline src/compiler.jl:1282]:
```suggestion
    otherMod, meta = GPUCompiler.codegen(:llvm, job; optimize=false, cleanup=false, validate=false, ctx)
```

File: Project.toml [hunk 2]
```diff
@@ -18,7 +18,7 @@ Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 [compat]
 Adapt = "3.3"
 CEnum = "0.4"
-Enzyme_jll = "0.0.33"
+Enzyme_jll = "0.0.35"
 GPUCompiler = "0.15, 0.16"
```
Comment (vchuravy) [inline Project.toml:22]:
```suggestion
GPUCompiler = "0.16.2"
```

**Comment by vchuravy** [comment]
Note to self: squash merge.

File: test/cuda.jl [hunk 3]
```diff
@@ -24,6 +24,50 @@ end
...
+    return nothing
+end
+
+@testset "cos_kernel" begin
+    A = CUDA.ones(64,)
+    @cuda threads=length(A) cos_kernel(A)
+    dA = similar(A)
+    dA .= 1
+    @cuda threads=length(A) grad_cos_kernel(A, dA)
+    @test_throws all(dA .== -sin(1))
```
Comment (vchuravy) [inline test/cuda.jl:68]:
```suggestion
    @test all(dA .== -sin(1))
```

File: test/cuda.jl [hunk 4]
```diff
@@ -24,6 +24,50 @@ end
...
+    return nothing
+end
+
+@testset "exp_kernel" begin
+    A = CUDA.ones(64,)
+    @cuda threads=length(A) exp_kernel(A)
+    dA = similar(A)
+    dA .= 1
+    @cuda threads=length(A) grad_exp_kernel(A, dA)
+    @test_throws all(dA .== exp(1))
```
Comment (vchuravy) [inline test/cuda.jl:46]:
```suggestion
    @test all(dA .== exp(1))
```

File: test/cuda.jl [hunk 5]
```diff
@@ -24,6 +24,50 @@ end
...
+function exp_kernel(A)
+    i = threadIdx().x
+    if i <= length(A)
+        A[i] = exp(A[i])
+    end
+    return nothing
+end
+
+function grad_exp_kernel(A, dA)
+    Enzyme.autodiff_deferred(cos_kernel, Const, Duplicated(A, dA))
```
Comment (vchuravy) [inline test/cuda.jl:36]:
```suggestion
    Enzyme.autodiff_deferred(exp_kernel, Const, Duplicated(A, dA))
```

**Pull request merged**